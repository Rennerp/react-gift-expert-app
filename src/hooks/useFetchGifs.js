import { useState, useEffect } from "react"
import { getGif } from "../helpers/getGifs";

// Esto es un custom hook, que realizara el trabajo del fetch a la API
export const useFetchGifs = (category) =>{
  //Cargamos el hoook setState
  const [state, setState] = useState({
    data: [],
    loading: true
  })
  // Utilizamos el useEffect para prevenir el trigger de la API
  // por cada cambio de estado del componente

  useEffect(() => {
    // Utilizamos el helper de getGiff para llamar a la API le enviamos la categoria
    // Nos regresa una promesa la trabajos con el .then
    getGif(category)
      .then( imgs => {
        // Cambiamos el estado del hook inicial
        setState({
          data: imgs,
          loading: false
        })
      })
  }, [ category ])

  // Realizamos un return del estado
  return state;
}