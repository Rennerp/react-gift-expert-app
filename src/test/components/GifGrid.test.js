import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Realizando pruebas del componente GifGrid', () => {
  const category = 'Samurai x'

  test('Debe de reenderizar el componente', () => {
    // Se realiza una falsa llamada con el estado inicial del hook
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    });

    const wrapper = shallow(<GifGrid category = { category } />);
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe de mostrar items cuando se carga una imagen desde el hook', () => {
    //Se utilizan mocks: son fakers de informacion

    const imgs = [{
      id: 'ABC',
      url: 'https://localhost/cualquier/cosa.jpg',
      title: 'Cualquier cosa'
    },{
      id: '123',
      url: 'https://localhost/cualquier/cosa.jpg',
      title: 'Cualquier cosa'
    }]

    useFetchGifs.mockReturnValue({
      data: imgs,
      loading: false
    });

    const wrapper = shallow(<GifGrid category = { category } />);
    // Verificamos que la etiqueta p no exista
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').exists()).toBe(false);
    expect(wrapper.find('GifGridItem').length).toBe(imgs.length);

  })

})

