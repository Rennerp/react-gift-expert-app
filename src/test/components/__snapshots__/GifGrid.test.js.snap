// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Realizando pruebas del componente GifGrid Debe de mostrar items cuando se carga una imagen desde el hook 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
     
    Samurai x
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Realizando pruebas del componente GifGrid Debe de reenderizar el componente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
     
    Samurai x
  </h3>
  <p>
     Loading 
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
